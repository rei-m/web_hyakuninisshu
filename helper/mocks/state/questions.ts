import { questionsTypes } from '@src/state/questions';

export const MOCK_YOMIFUDA_1: questionsTypes.YomiFuda = {
  karutaNo: 1,
  shoku: '秋の田の',
  niku: 'かりほの庵の',
  sanku: '苫をあらみ',
};

export const MOCK_TORIFUDA_1: questionsTypes.ToriFuda = {
  karutaNo: 1,
  shiku: 'きりにぬれつつ',
  kekku: 'わがころもでは',
};

export const MOCK_TORIFUDA_2: questionsTypes.ToriFuda = {
  karutaNo: 2,
  shiku: 'ころもほすてふ',
  kekku: 'あまのかぐやま',
};

export const MOCK_TORIFUDA_3: questionsTypes.ToriFuda = {
  karutaNo: 3,
  shiku: 'ながながしよを',
  kekku: 'ひとりかもねむ',
};

export const MOCK_TORIFUDA_4: questionsTypes.ToriFuda = {
  karutaNo: 4,
  shiku: 'ふじのたかねに',
  kekku: 'ゆきはふりつつ',
};
