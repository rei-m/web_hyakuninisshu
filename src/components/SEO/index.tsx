import * as React from 'react';
import Helmet from 'react-helmet';
import { graphql, StaticQuery } from 'gatsby';
import { SiteMetaData } from '@src/types';

export interface Props {
  description?: string;
  lang?: string;
  meta?: Array<{
    name: string;
    content: string;
  }>;
  keywords?: string[];
  title: string;
}

export interface DefaultSEOQueryData {
  site: {
    siteMetadata: SiteMetaData;
  };
  ogpImage: {
    publicURL: string;
  };
}

const SEO: React.FC<Props> = ({ description, lang, meta, keywords, title }) => {
  const checkedLang: string = lang ? lang : 'ja';
  const checkedKeywords: string[] = keywords ? keywords : [];
  const checkedMeta: Array<{
    name: string;
    content: string;
  }> = meta ? meta : [];

  return (
    <StaticQuery
      query={query}
      render={(data: DefaultSEOQueryData) => {
        const checkedTitle = title || data.site.siteMetadata.title;
        const checkedDescription = description || data.site.siteMetadata.description;
        return (
          <Helmet
            htmlAttributes={{
              lang: checkedLang,
            }}
            title={checkedTitle}
            meta={[
              {
                name: `description`,
                content: checkedDescription,
              },
              {
                property: `og:title`,
                content: checkedTitle,
              },
              {
                property: `og:description`,
                content: checkedDescription,
              },
              {
                property: `og:type`,
                content: `website`,
              },
              {
                property: `og:image`,
                content: `https://hyakuninanki.net${data.ogpImage.publicURL}`,
              },
              {
                name: `twitter:card`,
                content: `summary`,
              },
              {
                name: `twitter:creator`,
                content: data.site.siteMetadata.author,
              },
              {
                name: `twitter:title`,
                content: checkedTitle,
              },
              {
                name: `twitter:description`,
                content: checkedDescription,
              },
              {
                name: `twitter:image`,
                content: `https://hyakuninanki.net${data.ogpImage.publicURL}`,
              },
            ]
              .concat(
                checkedKeywords.length > 0
                  ? {
                      name: `keywords`,
                      content: checkedKeywords.join(`, `),
                    }
                  : []
              )
              .concat(checkedMeta)}
          />
        );
      }}
    />
  );
};

export default SEO;

const query = graphql`
  query DefaultSEOQuery {
    site {
      siteMetadata {
        title
        description
        author
      }
    }
    ogpImage: file(relativePath: { eq: "app-icon.png" }) {
      publicURL
    }
  }
`;
