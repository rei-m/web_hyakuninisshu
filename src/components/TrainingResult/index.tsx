import * as React from 'react';
import { Link } from 'react-router-dom';
// import styled from 'styled-components';

export interface TrainingResultOwnProps {
  readonly totalCount: number;
  readonly correctCount: number;
  readonly averageAnswerSecond: number;
}

export interface TrainingResultDispatchProps {
  readonly onClickRestart: () => void;
}

export type TrainingResultProps = TrainingResultOwnProps &
  TrainingResultDispatchProps;

const TrainingResult = ({
  averageAnswerSecond,
  correctCount,
  totalCount,
  onClickRestart
}: TrainingResultProps) => {
  return (
    <section>
      <div>
        <div>正解数</div>
        <div>
          {correctCount} / {totalCount}
        </div>
      </div>
      <div>
        <div>平均回答時間</div>
        <div>{averageAnswerSecond}秒</div>
      </div>
      {correctCount !== totalCount && (
        <button onClick={onClickRestart}>間違えた歌の練習</button>
      )}
      <Link to="/training" replace={true}>
        メニューに戻る
      </Link>
    </section>
  );
};

export default TrainingResult;
