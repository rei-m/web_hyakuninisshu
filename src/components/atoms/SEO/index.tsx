import React from 'react';
import Helmet from 'react-helmet';
import { useSEO } from '@src/hooks/staticQueries/useSEO';

export type Props = {
  description?: string;
  lang?: string;
  meta?: Array<{
    name: string;
    content: string;
  }>;
  keywords?: string[];
  title: string;
};

const SEO = ({ description, lang = 'ja', meta = [], keywords = [], title }: Props) => {
  const { site, ogpImage } = useSEO();
  const checkedTitle = title || site.siteMetadata.title;
  const checkedDescription = description || site.siteMetadata.description;
  return (
    <Helmet
      htmlAttributes={{
        lang,
      }}
      title={checkedTitle}
      meta={[
        {
          name: `description`,
          content: checkedDescription,
        },
        {
          property: `og:title`,
          content: checkedTitle,
        },
        {
          property: `og:description`,
          content: checkedDescription,
        },
        {
          property: `og:type`,
          content: `website`,
        },
        {
          property: `og:image`,
          content: `https://hyakuninanki.net${ogpImage.publicURL}`,
        },
        {
          name: `twitter:card`,
          content: `summary`,
        },
        {
          name: `twitter:creator`,
          content: site.siteMetadata.author,
        },
        {
          name: `twitter:title`,
          content: checkedTitle,
        },
        {
          name: `twitter:description`,
          content: checkedDescription,
        },
        {
          name: `twitter:image`,
          content: `https://hyakuninanki.net${ogpImage.publicURL}`,
        },
      ]
        .concat(
          keywords.length > 0
            ? {
                name: `keywords`,
                content: keywords.join(`, `),
              }
            : []
        )
        .concat(meta)}
    />
  );
};

export default SEO;
