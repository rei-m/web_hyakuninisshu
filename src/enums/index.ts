// このIFは苦肉の策だが他にいい方法はあるのか。。。
export interface EnumCollection<T> {
  readonly values: T[];
  readonly valueOf: (value: T) => T;
}

export enum MenuType {
  Training,
  Exam,
  Material,
  Other
}

export enum QuestionState {
  InAnswer,
  Answered,
  ConfirmCorrect,
  Finished
}

export enum RangeFromCondition {
  One = 1,
  Eleven = 11,
  TwentyOne = 21,
  ThirtyOne = 31,
  FortyOne = 41,
  FiftyOne = 51,
  SixtyOne = 61,
  SeventyOne = 71,
  EightyOne = 81,
  NinetyOne = 91
}

export const RangeFromConditions: EnumCollection<RangeFromCondition> = {
  valueOf: (value: number) => {
    const result = RangeFromConditions.values.find(v => v === value);
    if (result !== undefined) {
      return result;
    }
    throw new Error(`unknown RangeFrom, value is ${value}`);
  },
  values: [
    RangeFromCondition.One,
    RangeFromCondition.Eleven,
    RangeFromCondition.TwentyOne,
    RangeFromCondition.ThirtyOne,
    RangeFromCondition.FortyOne,
    RangeFromCondition.FiftyOne,
    RangeFromCondition.SixtyOne,
    RangeFromCondition.SeventyOne,
    RangeFromCondition.EightyOne,
    RangeFromCondition.NinetyOne
  ]
};

export enum RangeToCondition {
  Ten = 10,
  Twenty = 20,
  Thirty = 30,
  Forty = 40,
  Fifty = 50,
  Sixty = 60,
  Seventy = 70,
  Eighty = 80,
  Ninety = 90,
  OneHundred = 100
}

export const RangeToConditions: EnumCollection<RangeToCondition> = {
  valueOf: (value: number) => {
    const result = RangeToConditions.values.find(v => v === value);
    if (result !== undefined) {
      return result;
    }
    throw new Error(`unknown RangeTo, value is ${value}`);
  },
  values: [
    RangeToCondition.Ten,
    RangeToCondition.Twenty,
    RangeToCondition.Thirty,
    RangeToCondition.Forty,
    RangeToCondition.Fifty,
    RangeToCondition.Sixty,
    RangeToCondition.Seventy,
    RangeToCondition.Eighty,
    RangeToCondition.Ninety,
    RangeToCondition.OneHundred
  ]
};

export enum KimarijiCondition {
  None = 0,
  One = 1,
  Two = 2,
  Three = 3,
  Four = 4,
  Five = 5,
  Six = 6
}

export const KimarijiConditions: EnumCollection<KimarijiCondition> = {
  valueOf: (value: number) => {
    const result = KimarijiConditions.values.find(v => v === value);
    if (result !== undefined) {
      return result;
    }
    throw new Error(`unknown Kimariji, value is ${value}`);
  },
  values: [
    KimarijiCondition.None,
    KimarijiCondition.One,
    KimarijiCondition.Two,
    KimarijiCondition.Three,
    KimarijiCondition.Four,
    KimarijiCondition.Five,
    KimarijiCondition.Six
  ]
};

export enum ColorCondition {
  None = '',
  Blue = 'blue',
  Pink = 'pink',
  Yellow = 'yellow',
  Green = 'green',
  Orange = 'orange'
}

export const ColorConditions: EnumCollection<ColorCondition> = {
  valueOf: (value: string) => {
    const result = ColorConditions.values.find(v => v === value);
    if (result !== undefined) {
      return result;
    }
    throw new Error(`unknown Color, value is ${value}`);
  },
  values: [
    ColorCondition.None,
    ColorCondition.Blue,
    ColorCondition.Pink,
    ColorCondition.Yellow,
    ColorCondition.Green,
    ColorCondition.Orange
  ]
};

export enum KarutaStyleCondition {
  KanjiAndKana = 0,
  KanaOnly = 1
}

export const KarutaStyleConditions: EnumCollection<KarutaStyleCondition> = {
  valueOf: (value: number) => {
    const result = KarutaStyleConditions.values.find(v => v === value);
    if (result !== undefined) {
      return result;
    }
    throw new Error(`unknown KarutaStyle, value is ${value}`);
  },
  values: [KarutaStyleCondition.KanjiAndKana, KarutaStyleCondition.KanaOnly]
};
