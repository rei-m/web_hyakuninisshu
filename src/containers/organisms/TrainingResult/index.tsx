import React from 'react';
import { useSelector } from 'react-redux';
import KarutaPlayingResult from '@src/components/organisms/KarutaPlayingResult';
import { GlobalState } from '@src/state';
import { questionsTypes } from '@src/state/questions';
import { Answer } from '@src/types';

export interface OwnProps {
  onClickBack: () => void;
  onClickRestart: () => void;
}

export interface ConnectedProps {
  answers: Answer[];
}

export type PresenterProps = OwnProps & ConnectedProps;

export type ContainerProps = OwnProps & { presenter: React.FC<PresenterProps> };

export const TrainingResultPresenter = ({ answers, onClickBack, onClickRestart }: PresenterProps) => (
  <KarutaPlayingResult answers={answers} onClickBack={onClickBack} onClickRestart={onClickRestart} />
);

export const TrainingResultContainer = ({ presenter, onClickBack, onClickRestart }: ContainerProps) => {
  const { answers } = useSelector<GlobalState, questionsTypes.State>(state => state.questions);
  return presenter({ answers: answers ? answers : [], onClickBack, onClickRestart });
};

export const TrainingResult = (props: OwnProps) => (
  <TrainingResultContainer {...props} presenter={TrainingResultPresenter} />
);

export default TrainingResult;
