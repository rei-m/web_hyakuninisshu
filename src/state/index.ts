import { applyMiddleware, combineReducers, createStore as reduxCreateStore, Reducer, Store } from 'redux';
import thunk from 'redux-thunk';
import { questions, QuestionsState } from '@src/state/questions';
import { ui, UiState } from '@src/state/ui';

export interface GlobalState {
  questions: QuestionsState;
  ui: UiState;
}

export const reducer: Reducer<GlobalState> = combineReducers<GlobalState>({
  questions,
  ui,
});

export const createStore = (): Store<GlobalState> => reduxCreateStore(reducer, applyMiddleware(thunk));
