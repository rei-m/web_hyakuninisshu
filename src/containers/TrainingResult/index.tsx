import * as React from 'react';
import { connect } from 'react-redux';
import { ThunkDispatch } from 'redux-thunk';
import TrainingResultView, { Props as TrainingResultViewProps } from '@src/components/TrainingResultView';
import ErrorMessage from '@src/components/ErrorMessage';
import { GlobalState } from '@src/state';
import { restartQuestions, QuestionsActions } from '@src/actions/questions';
import { QuestionState } from '@src/enums';
import { Answer } from '@src/types';

export interface ConnectedProps {
  answers: Answer[];
  totalCount: number;
  questionState?: QuestionState;
}

export type DispatchProps = Pick<TrainingResultViewProps, 'onClickRestart'>;

export type Props = ConnectedProps & DispatchProps;

export const TrainingResult: React.FC<Props> = ({ answers, totalCount, questionState, onClickRestart }) => {
  if (questionState !== QuestionState.Finished) {
    return <ErrorMessage text="不正な遷移を行いました。前の画面からやり直してください。" />;
  }
  const averageAnswerSecond = answers.reduce((prev, current) => prev + current.time, 0) / 1000 / totalCount;
  return (
    <TrainingResultView
      averageAnswerSecond={Math.round(averageAnswerSecond * 100) / 100}
      totalCount={totalCount}
      correctCount={answers.filter(a => a.correct).length}
      onClickRestart={onClickRestart}
    />
  );
};

export const mapStateToProps = ({ questions }: GlobalState): ConnectedProps => {
  const { answers, questionState } = questions;
  return {
    answers,
    totalCount: questions.questions.length,
    questionState,
  };
};

export const mapDispatchToProps = (dispatch: ThunkDispatch<GlobalState, {}, QuestionsActions>): DispatchProps => ({
  onClickRestart: () => {
    dispatch(restartQuestions());
  },
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(TrainingResult);
