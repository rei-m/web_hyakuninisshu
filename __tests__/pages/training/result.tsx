import * as ReactTestRenderer from 'react-test-renderer';
import * as React from 'react';
import { Provider } from 'react-redux';
import TrainingResult, { QueryData as PageQueryData } from '@src/pages/training/result';
import { DefaultSEOQueryData } from '@src/components/atoms/SEO';
import { createStore } from '@src/state';
import { setUpQueryOnce } from '@test/helpers';
import { QuestionState } from '@src/enums';

describe('/training/result', () => {
  beforeEach(() => {
    setUpQueryOnce<DefaultSEOQueryData>({
      site: {
        siteMetadata: {
          title: 'exam result title',
          description: 'exam resul description',
          author: '@rei-m',
        },
      },
      ogpImage: {
        publicURL: 'http://localhost:8000/app.png',
      },
    });
    setUpQueryOnce<PageQueryData>({
      trainingResultBGImage: {
        publicURL: 'http://localhost:8000/app.png',
      },
    });
  });

  it('should render component', () => {
    const store = createStore();
    const questions = store.getState().questions;
    store.getState().questions = { ...questions, questionState: QuestionState.Finished };

    const renderer = ReactTestRenderer.create(
      <Provider store={store}>
        <TrainingResult />
      </Provider>
    );
    expect(renderer.toJSON()).toMatchSnapshot();
  });
});
