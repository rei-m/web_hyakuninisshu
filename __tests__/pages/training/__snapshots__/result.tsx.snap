// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/training/result should render component 1`] = `
Array [
  <header
    className="sc-bwzfXH geYBAS"
  >
    <div
      className="sc-bdVaJa flBAph"
      onMouseDown={[Function]}
      onMouseUp={[Function]}
    >
      <span
        className="sc-bxivhb iUaUQC"
        color="#fff"
        data-test="back"
        onClick={[Function]}
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root-1"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
          />
        </svg>
      </span>
    </div>
    <h1
      className="sc-htpNat eSnwWs"
    >
      百人一首 - 練習結果 -
    </h1>
  </header>,
  <div
    className="sc-gzVnrw eeAPtW"
  >
    <div
      className="sc-kGXeez HLrKz"
    >
      <div
        className="sc-iwsKbI hDfyDa"
      >
        <div
          className="sc-gZMcBi eXoTgE"
        >
          この欄は広告枠です
        </div>
      </div>
      <div
        className="sc-cSHVUG hirGEh"
      >
        <div
          className="sc-jTzLTM gkTcKL"
          style={
            Object {
              "marginBottom": "16px",
            }
          }
        >
          <div
            className="sc-fjdhpX jsSKD"
          >
            正解数
          </div>
          <div
            className="sc-jzJRlG fciurp"
          >
            0 / 0
          </div>
        </div>
        <div
          className="sc-jTzLTM gkTcKL"
          style={
            Object {
              "marginBottom": "32px",
            }
          }
        >
          <div
            className="sc-fjdhpX jsSKD"
          >
            平均回答時間
          </div>
          <div
            className="sc-jzJRlG fciurp"
          >
            NaN秒
          </div>
        </div>
        <div
          className="sc-kAzzGY ciXhiV"
        >
          <a
            className="twitter-share-button"
            data-show-count="false"
            data-size="large"
            data-text="百人一首で 0問中 0問 正解しました！"
            data-url="https://hyakuninanki.net"
            href="https://twitter.com/share"
          >
            Tweet
          </a>
        </div>
        <button
          className="MuiButtonBase-root-41 MuiButton-root-15 MuiButton-outlined-23 Component-button-10 Component-normalButton-11"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="0"
          type="button"
        >
          <span
            className="MuiButton-label-16"
          >
            <span
              aria-hidden="true"
              className="material-icons MuiIcon-root-44 MuiIcon-fontSizeSmall-51 Component-normalIcon-13"
            >
              arrow_back
            </span>
            メニューに戻る
          </span>
          <span
            className="MuiTouchRipple-root-53"
          />
        </button>
      </div>
      <div
        className="sc-gqjmRU dIdbVO"
      >
        <div
          className="sc-VigVT cYoXur"
        >
          この欄は広告枠です
        </div>
      </div>
    </div>
  </div>,
]
`;
